crates/oxc_span/src/cmp.rs

```diff
@@ -0,0 +1,158 @@
+    /// This method tests for contents of `self` and `other` to be equal.
+    #[must_use]
+    fn content_eq(&self, other: &Rhs) -> bool;
+
+    /// This method tests for contents of `self` and `other` not to be equal.
+    /// The default implementation is almost always
+    /// sufficient, and should not be overridden without very good reason.
+    #[inline]
+    #[must_use]
+    fn content_ne(&self, other: &Rhs) -> bool {
+        !self.content_eq(other)
+    }
+}
+
+impl ContentEq for () {
+    #[inline]
+    fn content_eq(&self, _other: &()) -> bool {
+        true
+    }
+
+    #[inline]
+    fn content_ne(&self, _other: &()) -> bool {
+        false
+    }
+}
+
+/// Blanket implementation for references
+impl<A: ?Sized, B: ?Sized> ContentEq<&B> for &A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mutable references
+impl<A: ?Sized, B: ?Sized> ContentEq<&mut B> for &mut A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&mut B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&mut B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mixed references
+impl<A: ?Sized, B: ?Sized> ContentEq<&mut B> for &A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&mut B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&mut B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for mixed references
+impl<A: ?Sized, B: ?Sized> ContentEq<&B> for &mut A
+where
+    A: ContentEq<B>,
+{
+    #[inline]
+    fn content_eq(&self, other: &&B) -> bool {
+        ContentEq::content_eq(*self, *other)
+    }
+    #[inline]
+    fn content_ne(&self, other: &&B) -> bool {
+        ContentEq::content_ne(*self, *other)
+    }
+}
+
+/// Blanket implementation for [Option] types
+impl<T: ContentEq> ContentEq for Option<T> {
+    #[inline]
+    fn content_eq(&self, other: &Self) -> bool {
+        // NOTE: based on the standard library
+        // Spelling out the cases explicitly optimizes better than
+        // `_ => false`
+        #[allow(clippy::match_same_arms)]
+        match (self, other) {
+            (Some(lhs), Some(rhs)) => lhs.content_eq(rhs),
+            (Some(_), None) => false,
+            (None, Some(_)) => false,
+            (None, None) => true,
+        }
+    }
+}
+
+/// Blanket implementation for [oxc_allocator::Box] types
+impl<'a, T: ContentEq> ContentEq for oxc_allocator::Box<'a, T> {
+    #[inline]
+    fn content_eq(&self, other: &Self) -> bool {
+        self.as_ref().content_eq(other.as_ref())
+    }
+}
```

Comment:
Is this required? Doesn't auto-deref handle this?

e.g. even without this blanket impl, doesn't this work?

```rs
let box1 = Box::new_in(NullLiteral { span: Span::new(0, 1) }, allocator);
let box2 = Box::new_in(NullLiteral { span: Span::new(1, 2) }, allocator);
let equal = box1.content_eq(&box2);
```

Reply:
This one is required to allow this syntax:

```rust
let box1 = &Box::new_in(NullLiteral { span: Span::new(0, 1) }, allocator);
let box2 = &Box::new_in(NullLiteral { span: Span::new(1, 2) }, allocator);
let equal = ContentEq::content_eq(box1, box2);
```

This is really common in enums so it is worth having it around.

We are moving towards this syntax to avoid accidental hiccups in refactors. #5462

Reply:
Thanks for explaining. That makes sense.